runtime/middleware/jwt.go

```diff
@@ -0,0 +1,95 @@
+package middleware
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/hypermodeinc/modus/runtime/config"
+	"github.com/hypermodeinc/modus/runtime/logger"
+	"github.com/hypermodeinc/modus/runtime/utils"
+
+	"github.com/golang-jwt/jwt/v5"
+)
+
+type JWTClaimsKey string
+
+const JWTClaims JWTClaimsKey = "jwt_claims"
+
+func HandleJWT(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		var ctx context.Context = r.Context()
+		tokenStr := r.Header.Get("Authorization")
+		tokenStr = strings.TrimPrefix(tokenStr, "Bearer ")
+
+		privKeysStr := os.Getenv("MODUS_RSA_PEMS")
+		if privKeysStr == "" {
+			if !config.IsDevEnvironment() || tokenStr == "" {
+				next.ServeHTTP(w, r)
+				return
+			}
+			token, _, err := new(jwt.Parser).ParseUnverified(tokenStr, jwt.MapClaims{})
+			if err != nil {
+				logger.Debug(r.Context()).Err(err).Msg("JWT parse error")
+				next.ServeHTTP(w, r)
+				return
+			}
+			if claims, ok := token.Claims.(jwt.MapClaims); ok {
+				ctx = AddClaimsToContext(ctx, claims)
+			}
+			next.ServeHTTP(w, r.WithContext(ctx))
+			return
+
+		}
+
+		var privKeysUnmarshalled map[string]string
+		err := json.Unmarshal([]byte(privKeysStr), &privKeysUnmarshalled)
+		if err != nil {
+			logger.Error(r.Context()).Err(err).Msg("JWT private keys unmarshalling error")
+			http.Error(w, "Internal server error", http.StatusInternalServerError)
+			return
+		}
+
+		var token *jwt.Token
+
+		for _, privKey := range privKeysUnmarshalled {
```

Comment:
I don't think you need to loop, since the key function gives you the parsed token as a parameter, there's probably a way to match it to the correct key rather than looping over them all.

Reply:
i dont think so. to match it in O(1) (without looping) it means you would be able to extrapolate the public key for a jwt from just the jwt token itself which kinda breaks all of cryptography. but once we add endpoints we can ask them to specify what authProvider theyre using and then we can do it in O(1)